### ISSUE TITLE
Weird uninitialized variable error for destructured members of type parameter constrained to homomorphic mapped type where member is declared optional

### ISSUE THREAD

**Issue by DanielRosenwasser** [Microsoft/TypeScript#20994]
Found in Formik by @weswigham who patiently worked with me to find the minimal repro.

Run the following with `strictNullChecks`

```ts
interface Options {
    a?: number | object;
    b: () => void;
}

class C<T extends Options> {
    foo!: { [P in keyof T]: T[P] }

    method() {
        let { a, b } = this.foo;
        !(a && b)
        a
        // TypeScript issues an error on 'a' above:
        // Variable 'a' is used before being assigned.'
    }
}
```

**Comment by rozzzly** [comment]
@DanielRosenwasser 

```typescript
    foo!: { [P in keyof T]: T[P] }
```

Is the `!: ` some new/obscure syntax I'm not aware of? Or just a typo?

**Comment by ahejlsberg** [comment]
@rozzzly See #20166.

**Comment by sandersn** [comment]
Should be fixed #20995
